File: Sources/Util/UIView+MP.m
//
// Created by Maarten Billemont on 2020-06-04.
// Copyright (c) 2020 Lyndir. All rights reserved.
//

#import "UIView+MP.h"
#import <objc/runtime.h>

@interface UIView(MP_Swift)

- (NSString *)describeWithShort:(BOOL)s;

@end

@implementation UIView(MP)

- (UIEdgeInsets)alignmentRectInsets {

    return UIEdgeInsetsMake(
            -self.alignmentRectOutsets.top, -self.alignmentRectOutsets.left,
            -self.alignmentRectOutsets.bottom, -self.alignmentRectOutsets.right );
}

- (void)setAlignmentRectInsets:(UIEdgeInsets)alignmentRectInsets {

    self.alignmentRectOutsets = UIEdgeInsetsMake(
            -alignmentRectInsets.top, -alignmentRectInsets.left,
            -alignmentRectInsets.bottom, -alignmentRectInsets.right );
}

- (UIEdgeInsets)alignmentRectOutsets {

    return [objc_getAssociatedObject( self, @selector( alignmentRectOutsets ) ) UIEdgeInsetsValue];
}

- (void)setAlignmentRectOutsets:(UIEdgeInsets)alignmentRectOutsets {

    objc_setAssociatedObject( self, @selector( alignmentRectOutsets ),
            [NSValue valueWithUIEdgeInsets:alignmentRectOutsets], OBJC_ASSOCIATION_RETAIN );
}

- (CGRect)alignmentRect {

    return [self alignmentRectForFrame:self.frame];
}

- (NSString *)accessibilityIdentifier {

    if ([NSThread currentThread].threadDictionary[@"accessibilityIdentifier"])
        return nil;

    @try {
        [NSThread currentThread].threadDictionary[@"accessibilityIdentifier"] = @YES;
        return [self describeWithShort:NO];
    }
    @finally {
        [NSThread currentThread].threadDictionary[@"accessibilityIdentifier"] = nil;
    }
}

@end

File: Sources/Util/UIView+MP.swift
//
// Created by Maarten Billemont on 2019-11-08.
// Copyright (c) 2019 Lyndir. All rights reserved.
//

import UIKit

//private var inAccessibilityIdentifier = false

public extension UIView {
    override var accessibilityLabel: String? {
        get {
//            inAccessibilityIdentifier = true
//            defer { inAccessibilityIdentifier = false }
            super.accessibilityLabel ?? describe(self)
        }
        set {
            super.accessibilityLabel = newValue
        }
    }

    var owner : (UIResponder, String)? {
        var nextResponder: UIResponder?
        while let nextResponder_ = nextResponder {
            if let property = nextResponder_.ivarWithValue( self ) {
                return (nextResponder_, property)
            }

            nextResponder = nextResponder_.next
        }

        return nil
    }

//- (NSString *)infoPathName {
//
//    UIResponder *parent = [self nextResponder]
//    if ([parent isKindOfClass:[UIView class]])
//        return strf( @"%@/%@", [(UIView *)parent infoPathName]?: @"", [self infoShortName])
//
//    return strf( @"%@/%@", describe( [parent class] )?: @"", [self infoShortName] )
//}

}
